mllib/src/main/scala/org/apache/spark/ml/feature/RobustScaler.scala

```diff
@@ -147,49 +146,40 @@ class RobustScaler (override val uid: String)
 
   override def fit(dataset: Dataset[_]): RobustScalerModel = {
     transformSchema(dataset.schema, logging = true)
-    val localRelativeError = $(relativeError)
 
-    val summaries = dataset.select($(inputCol)).rdd.map {
-      case Row(vec: Vector) => vec
-    }.mapPartitions { iter =>
-      var agg: Array[QuantileSummaries] = null
-      while (iter.hasNext) {
-        val vec = iter.next()
-        if (agg == null) {
-          agg = Array.fill(vec.size)(
-            new QuantileSummaries(QuantileSummaries.defaultCompressThreshold, localRelativeError))
-        }
-        require(vec.size == agg.length,
-          s"Number of dimensions must be ${agg.length} but got ${vec.size}")
-        var i = 0
-        while (i < vec.size) {
-          agg(i) = agg(i).insert(vec(i))
-          i += 1
-        }
-      }
-
-      if (agg == null) {
-        Iterator.empty
-      } else {
-        Iterator.single(agg.map(_.compress))
-      }
-    }.treeReduce { (agg1, agg2) =>
-      require(agg1.length == agg2.length)
-      var i = 0
-      while (i < agg1.length) {
-        agg1(i) = agg1(i).merge(agg2(i))
-        i += 1
-      }
-      agg1
+    val numFeatures = MetadataUtils.getNumFeatures(dataset, $(inputCol))
+    val vectors = dataset.select($(inputCol)).rdd.map {
+      case Row(vec: Vector) =>
+        require(vec.size == numFeatures,
+          s"Number of dimensions must be $numFeatures but got ${vec.size}")
+        vec
     }
 
-    val (range, median) = summaries.map { s =>
-      (s.query($(upper)).get - s.query($(lower)).get,
-        s.query(0.5).get)
-    }.unzip
-
-    copyValues(new RobustScalerModel(uid, Vectors.dense(range).compressed,
-      Vectors.dense(median).compressed).setParent(this))
+    val localRelativeError = $(relativeError)
+    val localUpper = $(upper)
+    val localLower = $(lower)
+
+    val (ranges, medians) = vectors.mapPartitions { iter =>
+      if (iter.hasNext) {
+        iter.flatMap { vec => Iterator.range(0, numFeatures).map(i => (i, vec(i))) } ++
```

Comment:
You'll effectively filter NaNs later anyway by not inserting them. Do you want to filter them explicitly up here to avoid extra compress calls? Would this change behavior if NaN is in the data? I'd guess you would get NaN or an error before, but would just ignore them now.

Reply:
Good point!
Existing impl do not take NaN into account, and a `QuantileSummaries` can absorb a NaN without throwing an exception, however it can not be used then:
```scala
scala> import org.apache.spark.sql.catalyst.util.QuantileSummaries
import org.apache.spark.sql.catalyst.util.QuantileSummaries

scala> val s1 = new QuantileSummaries(QuantileSummaries.defaultCompressThreshold, 0.0001)
s1: org.apache.spark.sql.catalyst.util.QuantileSummaries = org.apache.spark.sql.catalyst.util.QuantileSummaries@577dac16

scala> var s1 = new QuantileSummaries(QuantileSummaries.defaultCompressThreshold, 0.0001)
s1: org.apache.spark.sql.catalyst.util.QuantileSummaries = org.apache.spark.sql.catalyst.util.QuantileSummaries@5006a697

scala> s1 = s1.insert(Double.NaN)
s1: org.apache.spark.sql.catalyst.util.QuantileSummaries = org.apache.spark.sql.catalyst.util.QuantileSummaries@5006a697

scala> s1 = s1.compress
s1: org.apache.spark.sql.catalyst.util.QuantileSummaries = org.apache.spark.sql.catalyst.util.QuantileSummaries@5680f009

scala> s1.query(0.5)
res1: Option[Double] = Some(NaN)
```
after absorbing a NaN value, it return NaN in `qurey`.

Then I test sklearn and it just ignore NaN now, so I guess we can also skip NaN for now
```python
from sklearn.preprocessing import RobustScaler
import numpy as np

X = np.array([[0.0, 0.0], [np.nan, -0.5], [1.0, -1.0], [1.5, np.nan], [2.0, -2.0]])

transformer = RobustScaler().fit(X)

transformer.center_
Out[5]: array([ 1.25, -0.75])

transformer.scale_
Out[6]: array([0.875, 0.875])

transformer.transform(X)
Out[7]: 
array([[-1.42857143,  0.85714286],
       [        nan,  0.28571429],
       [-0.28571429, -0.28571429],
       [ 0.28571429,         nan],
       [ 0.85714286, -1.42857143]])
```

Reply:
OK, and this is new in 3.0 I suppose, so, no need for a release note. It might be worth docuemnting the behavior in scaladoc